keycloak/keycloak-operator
# INTLY-4129 Pod Monitor implementation

## JIRA ID

https://issues.redhat.com/browse/INTLY-4129

## Additional Information

This PR adds another monitor resource: PodMonitor. It is useful when there are multiple Keycloak pods running at the same time in a cluster (HA mode).

## Verification Steps

1. Checkout this branch `add-pod-monitor` on my fork `git@github.com:davidkirwan/keycloak-operator.git`
2. apply all the CRDs etc: `make cluster/prepare`
3. Deploy the operator locally: `operator-sdk up local --namespace "keycloak"`
4. Modify the Keycloak CR to increase the instances to `3`:
```
diff --git a/deploy/examples/keycloak/keycloak.yaml b/deploy/examples/keycloak/keycloak.yaml
index d5d9d24..5d23118 100644
--- a/deploy/examples/keycloak/keycloak.yaml
+++ b/deploy/examples/keycloak/keycloak.yaml
@@ -5,8 +5,8 @@ metadata:
   labels:
     app: sso
 spec:
-  instances: 1
+  instances: 3
   extensions:
     - https://github.com/aerogear/keycloak-metrics-spi/releases/download/1.0.4/keycloak-metrics-spi-1.0.4.jar
   externalAccess:
-    enabled: True
\ No newline at end of file
+    enabled: True
```
5. label the keycloak namespace. `oc label namespace keycloak monitoring-key=middleware`
6. Install application monitoring stack - https://github.com/integr8ly/application-monitoring-operator
7. Then apply the Keycloak CR: `oc apply -f deploy/examples/keycloak/keycloak.yaml`
8. Then apply the Keycloak Realm CR: `oc apply -f deploy/examples/realm/basic_realm.yaml`
9. Verify that the PodMonitor created has the `monitoring-key: middleware` label
10. Check for metrics being available related to keycloak in the prometheus instance created by the application monitoring operator, eg: `keycloak_request_duration_bucket`

Expect log messages such as the following:
```
{"level":"info","ts":1576850553.5705435,"logger":"controller_keycloak","msg":"Reconciling Keycloak","Request.Namespace":"keycloak","Request.Name":"example-keycloak"}
{"level":"info","ts":1576850553.6047091,"logger":"action_runner","msg":"(    0)    SUCCESS Update Keycloak admin secret"}
{"level":"info","ts":1576850553.640096,"logger":"action_runner","msg":"(    1)    SUCCESS update keycloak prometheus rule"}
{"level":"info","ts":1576850553.6730623,"logger":"action_runner","msg":"(    2)    SUCCESS update keycloak service monitor"}
{"level":"info","ts":1576850553.7066567,"logger":"action_runner","msg":"(    3)    SUCCESS update keycloak pod monitor"}
{"level":"info","ts":1576850553.7443936,"logger":"action_runner","msg":"(    4)    SUCCESS update keycloak grafana dashboard"}
{"level":"info","ts":1576850553.7768626,"logger":"action_runner","msg":"(    5)    SUCCESS Update Database Secret"}
{"level":"info","ts":1576850553.812751,"logger":"action_runner","msg":"(    6)    SUCCESS Update Postgresql PersistentVolumeClaim"}
{"level":"info","ts":1576850553.846172,"logger":"action_runner","msg":"(    7)    SUCCESS Update Postgresql Deployment"}
{"level":"info","ts":1576850553.8792517,"logger":"action_runner","msg":"(    8)    SUCCESS Update Postgresql KeycloakService"}
{"level":"info","ts":1576850553.9124768,"logger":"action_runner","msg":"(    9)    SUCCESS Update keycloak Service"}
{"level":"info","ts":1576850553.9456978,"logger":"action_runner","msg":"(   10)    SUCCESS Update keycloak Discovery Service"}
{"level":"info","ts":1576850553.9791303,"logger":"action_runner","msg":"(   11)    SUCCESS Update Keycloak Deployment (StatefulSet)"}
{"level":"info","ts":1576850554.0188434,"logger":"action_runner","msg":"(   12)    SUCCESS Update Keycloak Route"}
{"level":"info","ts":1576850554.0523052,"logger":"controller_keycloak","msg":"desired cluster state met"}
```

Should see a PodMonitor created in the `keycloak` namespace:
![Screenshot from 2019-12-20 14-04-58](https://user-images.githubusercontent.com/1700165/71260020-c6914400-2331-11ea-8950-1db8079587bc.png)

Should see metrics from the multiple keycloak pods:
![Screenshot from 2019-12-20 14-10-01](https://user-images.githubusercontent.com/1700165/71260295-6f3fa380-2332-11ea-9f60-3950e9936b90.png)


## Checklist:

## Additional Notes
